### PR TITLE
optimize LastSeenMessage

### PR DESCRIPTION
**Pull request by KickSeason**
Apply #1802 on `master`.

* Record LastSeenMessage precisely, especially after vote.

### CODE REVIEW

**Comment by erikzhang** [comment]
In NEO3, the count of consensus nodes is fixed. So this is not needed, isn't it?

**Comment by KickSeason** [comment]
> In NEO3, the count of consensus nodes is fixed. So this is not needed, isn't it?

What if validators changed. Still use previous record?

File: src/neo/Consensus/ConsensusContext.cs [hunk 1]
```diff
@@ -386,9 +386,21 @@ public void Reset(byte viewNumber)
...
+                if (0 < Snapshot.Height && Blockchain.Singleton.GetBlock(Snapshot.Height - 1).NextConsensus != Blockchain.Singleton.GetBlock(Snapshot.Height).NextConsensus)
+                {
+                    var previous_last_seen_message = LastSeenMessage;
+                    LastSeenMessage = new Dictionary<ECPoint, int>();
+                    foreach (var validator in Validators)
+                    {
+                        if (previous_last_seen_message.TryGetValue(validator, out int value))
+                            LastSeenMessage[validator] = value;
+                        else
+                            LastSeenMessage[validator] = -1;
```
Comment (erikzhang) [inline src/neo/Consensus/ConsensusContext.cs:402]: `-1` will be counted in `CountFailed`. Maybe we should set it to the current height.

Comment (KickSeason) [reply src/neo/Consensus/ConsensusContext.cs]: done.

File: src/neo/Consensus/ConsensusContext.cs [hunk 2]
```diff
@@ -386,9 +386,21 @@ public void Reset(byte viewNumber)
                 CommitPayloads = new ConsensusPayload[Validators.Length];
                 if (LastSeenMessage == null)
                 {
-                    LastSeenMessage = new int[Validators.Length];
+                    LastSeenMessage = new Dictionary<ECPoint, int>();
                     for (int i = 0; i < Validators.Length; i++)
-                        LastSeenMessage[i] = -1;
+                        LastSeenMessage[Validators[i]] = -1;
```
Comment (shargon) [inline src/neo/Consensus/ConsensusContext.cs:391]:
```suggestion
                        LastSeenMessage[Validators[i]] = (int)Snapshot.Height;
```

Comment (vncoelho) [reply src/neo/Consensus/ConsensusContext.cs:391]: makes sense.

File: src/neo/Consensus/ConsensusContext.cs [hunk 3]
```diff
@@ -386,9 +386,21 @@ public void Reset(byte viewNumber)
                 CommitPayloads = new ConsensusPayload[Validators.Length];
                 if (LastSeenMessage == null)
                 {
-                    LastSeenMessage = new int[Validators.Length];
+                    LastSeenMessage = new Dictionary<ECPoint, int>();
                     for (int i = 0; i < Validators.Length; i++)
-                        LastSeenMessage[i] = -1;
+                        LastSeenMessage[Validators[i]] = (int)Snapshot.Height;
+                }
+                if (0 < Snapshot.Height && Blockchain.Singleton.GetBlock(Snapshot.Height - 1).NextConsensus != Blockchain.Singleton.GetBlock(Snapshot.Height).NextConsensus)
```
Comment (vncoelho) [inline src/neo/Consensus/ConsensusContext.cs:393]: Since most of the time `0 < Snapshot.Height` is true, perhaps this check could be the second.

Comment (vncoelho) [inline src/neo/Consensus/ConsensusContext.cs:393]: `if (Blockchain.Singleton.GetBlock(Snapshot.Height - 1).NextConsensus != Blockchain.Singleton.GetBlock(Snapshot.Height).NextConsensus && 0 < Snapshot.Height)`

File: src/neo/Consensus/ConsensusContext.cs [hunk 4]
```diff
@@ -386,9 +386,21 @@ public void Reset(byte viewNumber)
...
+                {
+                    var previous_last_seen_message = LastSeenMessage;
```
Comment (vncoelho) [inline src/neo/Consensus/ConsensusContext.cs:395]: Maybe this part could be a function in order to leave ConsensusContext more clear.

File: src/neo/Consensus/ConsensusContext.cs (hunk 3 above)
Comment (KickSeason) [reply src/neo/Consensus/ConsensusContext.cs:393]: `Snapshot.Height` could be `0`.

File: src/neo/Consensus/ConsensusContext.cs (hunk 4 above)
Comment (KickSeason) [reply src/neo/Consensus/ConsensusContext.cs:396]: done.

File: src/neo/Consensus/ConsensusContext.cs (hunk 3 above)
Comment (KickSeason) [reply src/neo/Consensus/ConsensusContext.cs]: done.

Comment (KickSeason) [reply src/neo/Consensus/ConsensusContext.cs]: done.

**Comment by KickSeason** [comment]
👍

**Pull request merged**